llama-index-integrations/tools/llama-index-tools-cassandra/llama_index/tools/cassandra/cassandra_database_wrapper.py

```diff
@@ -0,0 +1,675 @@
+        """Execute a CQL query and return the results or an error message."""
+        try:
+            return self.run(query, fetch, include_columns, **kwargs)
+        except Exception as e:
+            """Format the error message"""
+            return f"Error: {e}\n{traceback.format_exc()}"
+
+    def get_keyspace_tables_str_no_throw(self, keyspace: str) -> str:
+        """Get the tables for the specified keyspace."""
+        try:
+            return self.get_keyspace_tables_str(keyspace)
+        except Exception as e:
+            """Format the error message"""
+            return f"Error: {e}\n{traceback.format_exc()}"
+
+    def get_keyspace_tables_str(self, keyspace: str) -> str:
+        """Get the tables for the specified keyspace."""
+        tables = self.get_keyspace_tables(keyspace)
+        schema_string = ""
+        for table in tables:
+            schema_string += table.as_markdown() + "\n\n"
+
+        return schema_string
+
+    def get_keyspace_tables(self, keyspace: str) -> List[Table]:
+        """Get the Table objects for the specified keyspace."""
+        schema = self._resolve_schema([keyspace])
+        if keyspace in schema:
+            return schema[keyspace]
+        else:
+            return []
+
+    def get_table_data_no_throw(
+        self, keyspace: str, table: str, predicate: str, limit: int
+    ) -> str:
+        """Get data from the specified table in the specified keyspace. Optionally can
+        take a predicate for the WHERE clause and a limit.
+        """
+        try:
+            return self.get_table_data(keyspace, table, predicate, limit)
+        except Exception as e:
+            """Format the error message"""
+            return f"Error: {e}\n{traceback.format_exc()}"
+
+    # This is a more basic string building function that doesn't use a query builder
+    # or prepared statements
+    # TODO: Refactor to use prepared statements
+    def get_table_data(
+        self, keyspace: str, table: str, predicate: str, limit: int
+    ) -> str:
+        """Get data from the specified table in the specified keyspace."""
+        query = f"SELECT * FROM {keyspace}.{table}"
+
+        if predicate:
+            query += f" WHERE {predicate}"
+        if limit:
+            query += f" LIMIT {limit}"
+
+        query += ";"
+
+        result = self.run(query, fetch="all")
+        return "\n".join(str(row) for row in result)
+
+    def get_context(self) -> Dict[str, Any]:
+        """Return db context that you may want in agent prompt."""
+        keyspaces = self._fetch_keyspaces()
+        return {"keyspaces": ", ".join(keyspaces)}
+
+    def format_keyspace_to_markdown(
+        self, keyspace: str, tables: Optional[List[Table]] = None
+    ) -> str:
+        """
+        Generates a markdown representation of the schema for a specific keyspace
+        by iterating over all tables within that keyspace and calling their
+        as_markdown method.
+
+        Parameters:
```

Comment:
Should use google-style docstring format

Reply:
Switched to Google Docstring format